lib/sqlite/sqlite.go

```diff
@@ -0,0 +1,165 @@
+func InitDb(
+	ctx context.Context,
+	name string,
+	db *sql.DB,
+	ddls []string,
+	versionMigrations []MigrationFunc,
+) error {
+
+	schemaVersion := len(versionMigrations) + 1
+
+	q, err := db.QueryContext(ctx, "SELECT name FROM sqlite_master WHERE type='table' AND name='_meta';")
+	if q != nil {
+		defer func() { _ = q.Close() }()
+	}
+
+	if errors.Is(err, sql.ErrNoRows) || !q.Next() {
+		// empty database, create the schema including the _meta table and its versions
+		ddls := append(metaDdl(uint64(schemaVersion)), ddls...)
+		for _, ddl := range ddls {
+			if _, err := db.Exec(ddl); err != nil {
+				return xerrors.Errorf("failed to %s database execute ddl %q: %w", name, ddl, err)
+			}
+		}
+		return nil
+	}
+
+	if err != nil {
+		return xerrors.Errorf("error looking for %s database _meta table: %w", name, err)
+	}
+
+	if err := q.Close(); err != nil {
+		return xerrors.Errorf("error closing %s database _meta table query: %w", name, err)
+	}
+
+	// check the schema version to see if we need to upgrade the database schema
+	var foundVersion int
+	if err = db.QueryRow("SELECT max(version) FROM _meta").Scan(&foundVersion); err != nil {
+		return xerrors.Errorf("invalid %s database version: no version found", name)
+	}
+
+	if foundVersion > schemaVersion {
+		return xerrors.Errorf("invalid %s database version: version %d is greater than the number of migrations %d", name, foundVersion, len(versionMigrations))
+	}
+
+	runVacuum := foundVersion != schemaVersion
+
+	for i := foundVersion + 1; i <= schemaVersion; i++ {
+		now := time.Now()
+
+		log.Infof("Migrating %s database to version %d...", name, i)
+
+		tx, err := db.BeginTx(ctx, nil)
+		if err != nil {
+			return xerrors.Errorf("failed to start %s database transaction: %w", name, err)
+		}
+		defer func() { _ = tx.Rollback() }()
+		// versions start at 1, but the migrations are 0-indexed where the first migration would take us to version 2
+		if err := versionMigrations[i-2](ctx, tx); err != nil {
+			return xerrors.Errorf("failed to migrate %s database to version %d: %w", name, i, err)
+		}
+		if _, err := tx.ExecContext(ctx, `INSERT OR IGNORE INTO _meta (version) VALUES (?)`, i); err != nil {
+			return xerrors.Errorf("failed to update %s database _meta table: %w", name, err)
+		}
+		if err := tx.Commit(); err != nil {
+			return xerrors.Errorf("failed to commit %s database v%d migration transaction: %w", name, i, err)
+		}
+
+		log.Infof("Successfully migrated %s database from version %d to %d in %s", name, i-1, i, time.Since(now))
+	}
+
+	if runVacuum {
+		// During the large migrations, we have likely increased the WAL size a lot, so lets do some
+		// simple DB administration to free up space (VACUUM followed by truncating the WAL file)
+		// as this would be a good time to do it when no other writes are happening.
+		log.Infof("Performing %s database vacuum and wal checkpointing to free up space after the migration", name)
+		_, err := db.ExecContext(ctx, "VACUUM")
```

Comment:
Is the "VACCUM" and "PRAGMA wal_checkpoint(TRUNCATE) stmt available in Postgress as well ?